azblob/azblob_service.go

```diff
@@ -0,0 +1,145 @@
+package azblob
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/streaming"
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob"
+	"github.com/pkg/errors"
+)
+
+const (
+	azureServiceURL = "https://%s.blob.core.windows.net/"
+)
+
+type Service struct {
+	Container   string
+	ServiceURL  string
+	AccountName string
+	AccountKey  string
+}
+
+func (s *Service) New() (azblob.ContainerClient, error) {
+	cred, err := azblob.NewSharedKeyCredential(s.AccountName, s.AccountKey)
+	if err != nil {
+		return azblob.ContainerClient{}, err
+	}
+	url := fmt.Sprintf(azureServiceURL, s.AccountName)
+
+	serviceClient, err := azblob.NewServiceClientWithSharedKey(url, cred, nil)
+	if err != nil {
+		return azblob.ContainerClient{}, err
+	}
+
+	return serviceClient.NewContainerClient(s.Container), nil
+
+}
+
+func (s *Service) Close() {
+}
+
+func (s *Service) ListBlobs(prefix, delimiter string) (blobs []string, err error) {
+	container, err := s.New()
+	if err != nil {
+		return []string{}, err
+	}
+	defer s.Close()
+
+	listOptions := &azblob.ContainerListBlobHierarchySegmentOptions{Prefix: &prefix}
+	pager := container.ListBlobsHierarchy(delimiter, listOptions)
+
+	for pager.NextPage(context.Background()) {
+		resp := pager.PageResponse()
+		for _, v := range resp.ContainerListBlobHierarchySegmentResult.Segment.BlobItems {
+			blobs = append(blobs, *v.Name)
+		}
+		for _, v := range resp.ContainerListBlobHierarchySegmentResult.Segment.BlobPrefixes {
+			blobs = append(blobs, *v.Name)
+		}
+	}
+
+	if err := pager.Err(); err != nil {
+		return []string{}, err
+	}
+
+	return blobs, nil
+}
+
+func (s *Service) GetBlobProperties(blob string) (azblob.GetBlobPropertiesResponse, error) {
+	container, err := s.New()
+	if err != nil {
+		return azblob.GetBlobPropertiesResponse{}, err
+	}
+	blobClient := container.NewBlockBlobClient(blob)
```

Comment:
nil: abstract this part to a function

```suggestion
func (s *Service) newBlockBlobClient() ... {
	....
}
```

Reply:
Addressed. Thanks